File: agents/issue-reconciler/issue_reconciler/api/fetchers/repo.py
from __future__ import annotations

from typing import TypedDict

from issue_reconciler.api.client import GitHubClient
from issue_reconciler.config import (
    MAX_TEXT_LENGTH,
    OPENSPEC_ARCHIVE_DIR,
    OPENSPEC_CHANGES_PATH,
    REPO_NAME,
    REPO_OWNER,
)

_TREE_QUERY = """
  query OpenSpecChangesTree($owner: String!, $name: String!, $expression: String!) {
    repository(owner: $owner, name: $name) {
      object(expression: $expression) {
        ... on Tree {
          entries { name type }
        }
      }
    }
  }
"""


class OpenSpecChange(TypedDict):
    name: str
    proposal_text: str


def _build_proposals_query(names: list[str]) -> str:
    aliases = "\n".join(
        f"""
    proposal_{i}: object(expression: "HEAD:{OPENSPEC_CHANGES_PATH}/{name}/proposal.md") {{
      ... on Blob {{ text }}
    }}"""
        for i, name in enumerate(names)
    )
    return f"""query OpenSpecProposals($owner: String!, $name: String!) {{
    repository(owner: $owner, name: $name) {{ {aliases} }}
  }}"""


def fetch_openspec_changes(client: GitHubClient) -> list[OpenSpecChange]:
    """Lists in-flight OpenSpec changes (excluding the archive) with their
    proposal text, truncated. Correlating a proposal to a specific issue
    number happens in the openspec_evidence spoke, not here - this fetcher
    only normalizes what the repo contains.
    """
    tree_data = client.query(
        _TREE_QUERY, {"owner": REPO_OWNER, "name": REPO_NAME, "expression": f"HEAD:{OPENSPEC_CHANGES_PATH}"}
    )
    obj = (tree_data.get("repository") or {}).get("object")
    dirs = [e for e in (obj or {}).get("entries", []) if e["type"] == "tree" and e["name"] != OPENSPEC_ARCHIVE_DIR]

    if not dirs:
        return []

    query = _build_proposals_query([d["name"] for d in dirs])
    proposals_data = client.query(query, {"owner": REPO_OWNER, "name": REPO_NAME})
    repository = proposals_data.get("repository") or {}

    return [
        {
            "name": d["name"],
            "proposal_text": ((repository.get(f"proposal_{i}") or {}).get("text") or "")[:MAX_TEXT_LENGTH],
        }
        for i, d in enumerate(dirs)
    ]

File: agents/issue-reconciler/api/fetchers/repo.ts
import type { GitHubClient } from '../client.ts';
import { MAX_TEXT_LENGTH, OPENSPEC_ARCHIVE_DIR, OPENSPEC_CHANGES_PATH, REPO_NAME, REPO_OWNER } from '../../config.ts';

export type OpenSpecChange = {
  name: string;
  proposalText: string;
};

const TREE_QUERY = `
  query OpenSpecChangesTree($owner: String!, $name: String!, $expression: String!) {
    repository(owner: $owner, name: $name) {
      object(expression: $expression) {
        ... on Tree {
          entries { name type }
        }
      }
    }
  }
`;

type TreeResponse = {
  repository: {
    object: { entries: Array<{ name: string; type: string }> } | null;
  } | null;
};

type ProposalsResponse = Record<string, { text: string } | null>;

function buildProposalsQuery(names: string[]): string {
  const aliases = names
    .map(
      (n, i) => `
    proposal_${i}: object(expression: "HEAD:${OPENSPEC_CHANGES_PATH}/${n}/proposal.md") {
      ... on Blob { text }
    }`,
    )
    .join('\n');

  return `query OpenSpecProposals($owner: String!, $name: String!) {
    repository(owner: $owner, name: $name) { ${aliases} }
  }`;
}

/**
 * Lists in-flight OpenSpec changes (excluding the archive) with their
 * proposal text, truncated. Correlating a proposal to a specific issue
 * number happens in the openspec-evidence spoke, not here - this fetcher
 * only normalizes what the repo contains.
 */
export async function fetchOpenSpecChanges(client: GitHubClient): Promise<OpenSpecChange[]> {
  const treeData = await client.query<TreeResponse>(TREE_QUERY, {
    owner: REPO_OWNER,
    name: REPO_NAME,
    expression: `HEAD:${OPENSPEC_CHANGES_PATH}`,
  });

  const dirs = (treeData.repository?.object?.entries ?? []).filter(
    (entry) => entry.type === 'tree' && entry.name !== OPENSPEC_ARCHIVE_DIR,
  );

  if (dirs.length === 0) return [];

  const query = buildProposalsQuery(dirs.map((d) => d.name));
  const proposalsData = await client.query<{ repository: ProposalsResponse | null }>(query, {
    owner: REPO_OWNER,
    name: REPO_NAME,
  });

  const repository = proposalsData.repository ?? {};

  return dirs.map((dir, i) => ({
    name: dir.name,
    proposalText: (repository[`proposal_${i}`]?.text ?? '').slice(0, MAX_TEXT_LENGTH),
  }));
}
